benchalerts/clients.py

```diff
@@ -0,0 +1,198 @@
+
+    base_url: str
+    timeout_s = 10
+
+    def __init__(self):
+        retry_strategy = Retry(
+            total=5,
+            status_forcelist=frozenset((429, 502, 503, 504)),
+            backoff_factor=4,  # will retry in 2, 4, 8, 16, 32 seconds
+        )
+        if os.getenv("UNIT_TESTING"):
+            from tests.mocks import MockAdapter
+
+            adapter = MockAdapter()
+        else:
+            adapter = HTTPAdapter(max_retries=retry_strategy)
+
+        self.session = requests.Session()
+        self.session.mount("https://", adapter)
+
+    def get(self, path: str) -> dict:
+        log.debug(f"GET {self.base_url}{path}")
+        res = self.session.get(url=self.base_url + path, timeout=self.timeout_s)
+        res.raise_for_status()
+        return res.json()
+
+    def post(self, path: str, json: dict) -> Optional[dict]:
+        log.debug(f"POST {self.base_url}{path} {json}")
+        res = self.session.post(
+            url=self.base_url + path, json=json, timeout=self.timeout_s
+        )
+        res.raise_for_status()
+        if res.content:
+            return res.json()
+
+
+class GithubRepoClient(_BaseClient):
+    """A client to interact with a Github repo.
+
+    Parameters
+    ----------
+    repo
+        The repo name, in the form 'owner/repo'.
+
+    Environment variables
+    ---------------------
+    GITHUB_API_TOKEN
+        A Github API token with ``repo`` access.
+    """
+
+    def __init__(self, repo: str):
+        token = os.getenv("GITHUB_API_TOKEN")
+        if not token:
+            fatal_and_log("Environment variable GITHUB_API_TOKEN not found")
+
+        super().__init__()
+        self.session.headers = {"Authorization": f"Bearer {token}"}
+        self.base_url = f"https://api.github.com/repos/{repo}"
+
+    def create_pull_request_comment(
+        self,
+        comment: str,
+        *,
+        pull_number: Optional[int] = None,
+        commit_sha: Optional[str] = None,
+    ):
+        """Create a comment on a pull request, specified either by pull request number
+        or commit SHA.
+
+        Parameters
+        ----------
+        comment
+            The comment text.
+        pull_number
+            The number of the pull request. Specify either this or ``commit_sha``.
+        commit_sha
+            The SHA of a commit associated with the pull request. Specify either this
+            or ``pull_number``.
+        """
+        if commit_sha:
+            pull_numbers = [
+                pull["number"] for pull in self.get(f"/commits/{commit_sha}/pulls")
+            ]
+            if len(pull_numbers) != 1:
+                fatal_and_log(
+                    "Need exactly 1 pull request associated with commit "
+                    f"'{commit_sha}'. Found {pull_numbers}."
+                )
+            pull_number = pull_numbers[0]
+        elif not pull_number:
+            fatal_and_log("Either specify pull_number or commit_sha")
```

Comment:
Could we rearrange this so each logical fork is in a clear block (or internal method, if you like)? Right now it's kind of weird that the failure case is in the middle of the other two cases; I'd expect it to be first or last. 

Should probably also fail or warn if both `pull_number` and `commit_sha` are specified instead of silently ignoring `pull_number`. Or check it matches that of `commit_sha` if you like?

Reply:
Sure, I'll try to organize this section better!

Since this package is intended to be used in automated scripts 100% of the time, I thought it unnecessary to check whether someone passed both SHA and number, since that would just be a failure to read the docs. The checking code would add a bit more complexity here too. But if you feel that might still be necessary I'm happy to add it.

I did keep the check whether both are missing, because I've seen scripts that do something like `create_pull_request_comment(commit_sha=os.getenv("COMMIT_SHA"))` and that env var could be missing due to an upstream failure.

Reply:
I just moved the fail block to the beginning; let me know if that's sufficient